[libc++][test] Fix more MSVC and Clang warnings

StephanTLavavej opened the pull request:
Found while running libc++'s tests with MSVC's STL.

As usual, I'm trying to strike a balance between avoiding "grab bag" PRs and creating a trillion PRs. After gathering the `static_cast` warning fixes in #74962, I believe that these changes are small enough to group together. Please let me know if you'd like me to decompose this further.

* `libcxx/test/std/algorithms/alg.sorting/alg.heap.operations/sort.heap/ranges_sort_heap.pass.cpp`
  + Fix Clang `-Wunused-variable`, because `LIBCPP_ASSERT` expands to nothing for MSVC's STL.
  + This is the same "always void-cast" change that #73437 applied to the neighboring `complexity.pass.cpp`. I missed that `ranges_sort_heap.pass.cpp` was also affected because we had disabled this test.
* `libcxx/test/std/input.output/file.streams/fstreams/ifstream.members/buffered_reads.pass.cpp`
* `libcxx/test/std/input.output/file.streams/fstreams/ofstream.members/buffered_writes.pass.cpp`
  + Fix MSVC "warning C4244: '`=`': conversion from '`__int64`' to '`_Ty`', possible loss of data".
  + This is a valid warning, possibly the best one that MSVC found in this entire saga. We're accumulating a `std::vector<std::streamsize>` and storing the result in `std::streamsize total_size` but we actually have to start with `std::streamsize{0}` or we'll truncate.
* `libcxx/test/std/input.output/filesystems/fs.enum/enum.path.format.pass.cpp`
  + Fix Clang `-Wunused-local-typedef` because the following usage is libc++-only.
  + I'm just expanding it at the point of use, and using the dedicated `LIBCPP_STATIC_ASSERT` to keep the line length down.
* `libcxx/test/std/input.output/syncstream/syncbuf/syncstream.syncbuf.assign/swap.pass.cpp`
  + Fix MSVC "warning C4242: 'argument': conversion from '`int`' to '`const _Elem`', possible loss of data".
  + This is a valid warning (possibly the second-best) as `sputc()` returns `int_type`. If `sputc()` returns something unexpected, we want to know, so we should separately say `expected.push_back(CharT('B'))`.
* `libcxx/test/std/language.support/support.dynamic/new.delete/new.delete.single/new.size_align_nothrow.pass.cpp`
* `libcxx/test/std/language.support/support.dynamic/new.delete/new.delete.single/new.size_nothrow.pass.cpp`
  + Fix MSVC "warning C6001: Using uninitialized memory '`x`'."
  + [N4964](https://wg21.link/N4964) \[new.delete.single\]/12:
    > *Effects:* The deallocation functions (\[basic.stc.dynamic.deallocation\]) called by a *delete-expression* (\[expr.delete\]) to render the value of `ptr` invalid.
  + \[basic.stc.general\]/4:
    > When the end of the duration of a region of storage is reached, the values of all pointers representing the address of any part of that region of storage become invalid pointer values (\[basic.compound\]). Indirection through an invalid pointer value and passing an invalid pointer value to a deallocation function have undefined behavior. Any other use of an invalid pointer value has implementation-defined behavior.
  + In certain configurations, after `delete x;` MSVC will consider `x` to be radioactive (and in other configurations, it'll physically null out `x` as a safety measure). We can copy it into `old_x` before deletion, which the implementation finds acceptable.
* `libcxx/test/std/ranges/range.adaptors/range.elements/general.pass.cpp`
* `libcxx/test/std/ranges/range.adaptors/range.elements/iterator/deref.pass.cpp`
  + Fix MSVC "warning C4242: 'initializing': conversion from '`_Ty`' to '`_Ty`', possible loss of data".
  + This was being emitted in `pair` and `tuple`'s perfect forwarding constructors. Passing `short{1}` allows MSVC to see that no truncation is happening.
* `libcxx/test/std/ranges/range.adaptors/range.elements/iterator/member_types.compile.pass.cpp`
  + Fix MSVC "warning C4242: 'initializing': conversion from '`_Ty`' to '`_Ty2`', possible loss of data".
  + Similarly, this was being emitted in `pair`'s perfect forwarding constructor. After passing `short{1}`, I reduced repetition by relying on CTAD. (I can undo that cleanup if it's stylistically undesirable.)
* `libcxx/test/std/utilities/function.objects/refwrap/refwrap.const/type_conv_ctor.pass.cpp`
  + Fix MSVC "warning C4930: '`std::reference_wrapper<int> purr(void)`': prototyped function not called (was a variable definition intended?)".
  + There's no reason for `purr()` to be locally declared (aside from isolating it to a narrow scope, which has minimal benefits); it can be declared like `meow()` above. :smile_cat:
* `libcxx/test/std/utilities/memory/util.smartptr/util.smartptr.shared/util.smartptr.shared.create/make_shared_for_overwrite.pass.cpp`
* `libcxx/test/std/utilities/smartptr/unique.ptr/unique.ptr.create/make_unique_for_overwrite.default_init.pass.cpp`
  + Fix MSVC static analysis warnings when replacing `operator new`:
    ```
    warning C28196: The requirement that '(_Param_(1)>0)?(return!=0):(1)' is not satisfied. (The expression does not evaluate to true.)
    warning C6387: 'return' could be '0':  this does not adhere to the specification for the function 'new'.
    warning C6011: Dereferencing NULL pointer 'reinterpret_cast<char *>ptr+i'.
    ```
  + All we need is a null check, which appears in other `operator new` replacements: https://github.com/llvm/llvm-project/blob/b85f1f9b182234ba366d78ae2174a149e44d08c1/libcxx/test/std/language.support/support.dynamic/new.delete/new.delete.single/new.size.replace.pass.cpp#L27-L28

StephanTLavavej committed:
Fix MSVC "warning C4244: '=': conversion from '__int64' to '_Ty', possible loss of data".

This is a valid warning. We're accumulating a `std::vector<std::streamsize>` and storing the result in `std::streamsize total_size` but we actually have to start with `std::streamsize{0}` or we'll truncate.

StephanTLavavej committed:
Fix MSVC "warning C4242: 'argument': conversion from 'int' to 'const _Elem', possible loss of data".

This is a valid warning as sputc() returns int_type. If sputc() returns something unexpected, we want to know, so we should separately say expected.push_back(CharT('B')).

StephanTLavavej committed:
Fix MSVC "warning C6001: Using uninitialized memory 'x'."

This warning is valid-ish. N4964 [new.delete.single]/12: "Effects: The deallocation functions (6.7.5.5.3) called by a delete-expression (7.6.2.9) to render the value of ptr invalid."
[basic.stc.general]/4: "When the end of the duration of a region of storage is reached, the values of all pointers representing the address of any part of that region of storage become invalid pointer values (6.8.4). Indirection through an invalid pointer value and passing an invalid pointer value to a deallocation function have undefined behavior. Any other use of an invalid pointer value has implementation-defined behavior."

In certain configurations, after `delete x;` MSVC will consider `x` to be radioactive (and in other configurations, it'll physically null out `x`). We can copy it into `old_x` before deletion, which the implementation finds acceptable.

StephanTLavavej committed:
Fix MSVC static analysis warnings when replacing `operator new`.

This emitted a bunch of warnings:

make_shared_for_overwrite.pass.cpp(61) : warning C28196: The requirement that '(_Param_(1)>0)?(return!=0):(1)' is not satisfied. (The expression does not evaluate to true.): Lines: 62, 63, 64, 63, 66
make_shared_for_overwrite.pass.cpp(61) : warning C6387: 'return' could be '0':  this does not adhere to the specification for the function 'new'. See line 64 for an earlier location where this can occur: Lines: 62, 63, 64, 63, 66, 61
make_shared_for_overwrite.pass.cpp(64) : warning C6011: Dereferencing NULL pointer 'reinterpret_cast<char *>ptr+i'. : Lines: 62, 63, 64

All we need is a null check, which appears in other `operator new` replacements.

StephanTLavavej committed:
Fix MSVC "warning C4242: 'initializing': conversion from '_Ty' to '_Ty2', possible loss of data".

This was being emitted in pair's perfect forwarding constructor. It's simple to start with the desired types and rely on CTAD.

StephanTLavavej committed:
Fix MSVC "warning C4242: 'initializing': conversion from '_Ty' to '_Ty', possible loss of data".

This was being emitted in pair and tuple's perfect forwarding constructors.

StephanTLavavej committed:
Fix -Wunused-variable in ranges_sort_heap.pass.cpp. Same "always void-cast" change as LLVM-73437.

StephanTLavavej committed:
Fix MSVC "warning C4930: 'std::reference_wrapper<int> purr(void)': prototyped function not called (was a variable definition intended?)".

There's no reason for purr() to be locally declared (aside from isolating it to a narrow scope); it can be declared like meow() above.

StephanTLavavej committed:
Fix -Wunused-local-typedef warning.

I'm just expanding it at the point of use, and using the dedicated LIBCPP_STATIC_ASSERT to keep the line length down.

StephanTLavavej committed:
Apply clang-format from CI.

mordante commented on libcxx/test/std/utilities/memory/util.smartptr/util.smartptr.shared/util.smartptr.shared.create/make_shared_for_overwrite.pass.cpp:
```diff
@@ -60,6 +60,9 @@ constexpr char pattern = 0xDE;
 void* operator new(std::size_t count) {
   void* ptr = std::malloc(count);
+  if (!ptr) {
+    std::abort(); // placate MSVC's unchecked malloc warning
```
I like the comment, but when not an assert instead?

StephanTLavavej replied:
Unfortunately, static analysis doesn't consider `assert`'s effect on control flow. I'm not sure if this is a compiler bug or by design (because `assert` is controlled by `NDEBUG`). That's why the codebase currently uses `if-abort`.

mordante commented on libcxx/test/std/utilities/function.objects/refwrap/refwrap.const/type_conv_ctor.pass.cpp:
```diff
@@ -57,10 +59,7 @@ int main(int, char**)
-    extern std::reference_wrapper<int> purr();
-    ASSERT_SAME_TYPE(decltype(true ? purr() : 0), int);
-  }
+  { ASSERT_SAME_TYPE(decltype(true ? purr() : 0), int); }
```
```suggestion
  ASSERT_SAME_TYPE(decltype(true ? purr() : 0), int);
```
Maybe the same for meow while you're at it.

StephanTLavavej replied:
Can do!

mordante commented on libcxx/test/std/utilities/smartptr/unique.ptr/unique.ptr.create/make_unique_for_overwrite.default_init.pass.cpp:
```diff
@@ -27,6 +27,9 @@ constexpr char pattern = 0xDE;
 void* operator new(std::size_t count) {
   void* ptr = std::malloc(count);
+  if (!ptr) {
+    std::abort(); // placate MSVC's unchecked malloc warning
```
assert?

StephanTLavavej replied:
Same reason, `assert(ptr);` doesn't silence the warning.

mordante replied:
Maybe mention that in the comment, to avoid somebody to beautify this code in the future.

StephanTLavavej replied:
And braces, so they all look the same and clang-format is happy.

mordante reviewed (approved):
Thanks for the patch. LGTM modulo some comments.

StephanTLavavej committed:
Scope and formatting cleanups for type_conv_ctor.pass.cpp.

clang-format the file.

Remove braces around the `meow` and `ASSERT_SAME_TYPE` lines,
as we don't need to isolate anything by introducing scopes.

Add braces to cleanly separate `i` and `ri` from `j` and `rj`,
as there is always a low risk of unintentional mixups with such names.

StephanTLavavej committed:
Update comments to add "(assert() won't silence it)".

StephanTLavavej committed:
Apply clang-format from CI and add braces.

StephanTLavavej commented:
Thanks! The checks are green (with only the usual sporadic timeouts/failures and recently fixed failure, plus stage3 generic-msan queued for over 6 hours) after addressing your comments with very simple commits, so I'll go ahead and merge this.
